ledger_device_sdk/src/tlv/tlv_generic.rs

```diff
@@ -0,0 +1,267 @@
+        pub tag_to_flag: fn(Tag) -> u64,
+    }
+
+    /// Received tags tracker implementation
+    impl Received {
+        /// Create a new Received tracker
+        pub const fn new(map: fn(Tag) -> u64) -> Self {
+            Self {
+                flags: 0,
+                tag_to_flag: map,
+            }
+        }
+        /// Reset the received tags tracker
+        pub fn reset(&mut self) {
+            self.flags = 0;
+        }
+    }
+
+    /// Set a unique tag as received
+    #[inline]
+    fn set_unique(received: &mut Received, tag: Tag) -> Result<()> {
+        let f = (received.tag_to_flag)(tag);
+        if f == 0 {
+            return Ok(());
+        }
+        if (received.flags & f) != 0 {
+            return Err(TlvError::DuplicateUniqueTag);
+        }
+        received.flags |= f;
+        Ok(())
+    }
+
+    /// Decode a DER-encoded unsigned integer
+    fn der_u32(input: &[u8], off: &mut usize) -> Result<u32> {
+        if *off >= input.len() {
+            return Err(TlvError::UnexpectedEof);
+        }
+        let b0 = input[*off];
+        *off += 1;
+        if b0 & 0x80 == 0 {
+            return Ok((b0 & 0x7F) as u32);
+        }
+        let n = (b0 & 0x7F) as usize;
+        if n == 0 || n > 4 {
+            return Err(TlvError::InvalidDerLength);
+        }
+        if *off + n > input.len() {
+            return Err(TlvError::UnexpectedEof);
+        }
+        let mut v: u32 = 0;
+        for _ in 0..n {
+            v = (v << 8) | input[*off] as u32;
+            *off += 1;
+        }
+        Ok(v)
+    }
+
+    /// TLV parsing configuration
+    pub struct ParseCfg<'a, O> {
+        /// Handlers
+        handlers: &'a [Handler<O>],
+        /// Common handler (called before specific)
+        pub common: Option<HandlerFn<O>>,
+    }
+
+    /// ParseCfg implementation
+    impl<'a, O> ParseCfg<'a, O> {
+        /// Create a new ParseCfg
+        pub const fn new(handlers: &'a [Handler<O>]) -> Self {
+            Self {
+                handlers,
+                common: None,
+            }
+        }
+    }
+
+    /// Parse TLV-encoded data
+    pub fn parse<'a, O>(
+        cfg: &ParseCfg<'a, O>,
+        payload: &'a [u8],
+        tlv_out: &mut O,
+        received: &mut Received,
+    ) -> Result<()> {
+        received.reset();
+
+        let mut off = 0usize;
+        while off < payload.len() {
+            let tag_start = off;
+            let tag = der_u32(payload, &mut off)?;
+            let len = der_u32(payload, &mut off)? as usize;
+            if off + len > payload.len() {
+                return Err(TlvError::UnexpectedEof);
+            }
+            let val = &payload[off..off + len];
+            let raw = &payload[tag_start..off + len];
+            off += len;
+            let data = TlvData {
+                tag,
+                value: val,
+                raw,
+            };
+            if let Some(f) = cfg.common {
+                if !f(&data, tlv_out)? {
```

Comment:
Maybe a small comment, branching is not trivial

Reply:
More specifically the fact that the boolean handles flow control is not very clear

Reply:
Done